edu_edfi_airflow/dags/earthbeam_dag.py

```diff
@@ -78,6 +79,81 @@ def build_local_raw_dir(self, tenant_code: str, api_year: int, grain_update: Opt
         )
         return raw_dir
 
+    @staticmethod
+    def partition_on_tenant_and_year(
+        csv_paths: Union[Union[str, Path], List[Union[str, Path]]],
+        output_dir: str,
+        tenant_col: str = "tenant_code",
+        tenant_map: dict = None,
+        year_col: str = "api_year",
+        year_map: dict = None,
+    ):
+        """
+        Preprocessing function to shard data to parquet on disk
+
+        :param csv_paths: one or more complete file paths pointing to input data
+        :param output_dir: root directory of the parquet
+        :param tenant_col: (optional) name of the column to use as tenant code
+        :param tenant_map: (optional) map values from the contents of tenant_col to valid tenant codes
+        :param year_col: (optional) name of the column to use as API year
+        :param year_map: (optional) map values from the contents of api_col to valid API years
+
+        :return:
+        """
+        # (expensive) imports here so that the airflow scheduler doesn't have to deal with them
+        import dask
+        import dask.dataframe as dd
+        import pandas as pd
+
+        tenant_code = "tenant_code"
+        api_year = "api_year"
+
+        input_is_list = True if isinstance(csv_paths, list) else False
+
+        final_csv_paths = []
+        if input_is_list:
+            final_csv_paths = csv_paths
+        else:
+            # otherwise we have a string, so wrap it in a list
+            final_csv_paths = [csv_paths]
+
+        for csv_path in final_csv_paths:
+            if not Path(csv_path).is_file() or not Path(csv_path).suffix == '.csv':
+                raise ValueError(f"Input path '{csv_path}' is not a path to a file whose name ends with '.csv'")
+
+        path_mapping = {
+            # use the input file basenames as the parquet directory names
+            csv_path: PurePath(output_dir, PurePath(csv_path).name).with_suffix('')
+            for csv_path in final_csv_paths
+        }
+
+        Path(output_dir).mkdir(exist_ok=True)
+
+        with dask.config.set({
+            "temporary_directory": "./dask_temp/",
+            "dataframe.convert-string": True,
+            }), pd.option_context("mode.string_storage", "pyarrow"):
+
+            for csv_path, parquet_path in path_mapping.items():
+                df = dd.read_csv(csv_path, dtype=str, na_filter=False).fillna('')
+
+                if tenant_col not in df:
+                    raise KeyError(f"provided tenant_code column '{tenant_col}' not present in data")
+                if year_col not in df:
+                    raise KeyError(f"provided api_year column '{year_col}' not present in data")    
+            
+                # if needed, add tenant_code and api_year as columns so we can partition on them
+                if tenant_map is not None:
```

Comment:
FYI that you can simplify this to the following:
```
df[tenant_code] = df[tenant_col].map(tenant_map, meta=dd.utils.make_meta(df[tenant_col]))
```
This won't affect performance at all, but it would remove a lambda. Same goes for API year.